Allow Opening BIQ Files

QuintillusCFC opened the pull request:
1. Go to Load Game
2. Navigate to the scenarios folder
3. Choose a BIQ file, then click Open
4. It loads in-game.

Known issues: For some reason all pine forests are snow-covered.

Recommended to merge https://github.com/C7-Game/Prototype/pull/82 first, as it fixes non-square maps, which is a lot of BIQs.

QuintillusCFC committed:
Add the ability to open BIQ files.

This is likely limited to BIQ files with custom rules and custom maps
at this point.  The same amount of data is imported as for SAVs -
terrains, tiles, and map size.

Some of this, but not all, can be combined across BIQs and SAVs.  E.g.
terrains types can be combined, but tiles cannot be (e.g. overlays and
terraforms may have changed).

QuintillusCFC committed:
Remove comment that is fixed in another branch.

WildWeazel commented:
This is largely duplicated from ImportSav. Can we extract the logical blocks to common methods? eg. InitGameData, ReadBytes, ReadTerrain, ReadTiles...

QuintillusCFC commented:
Good point, and largely yes.  The major exception being tiles, since there are separate SAV and BIQ tile objects whose data may be different - although even there, we can standardize the variable names and structure more than they are currently.

QuintillusCFC committed:
Refactor out shared code from BIQs and SAVs.

This is not a whole lot right now, but as we import more sections of the BIQ, it will become a higher proportion of the whole.

QuintillusCFC committed:
Fix all forests being Tundra Forests in BIQ files.  Due to an error in setting the base terrain.

Fix the name on ImportCiv3TerrainTypes and move it to the end instead of sandwiched between ImportSav and ImportBIQ.

QuintillusCFC committed:
Wire up the Load Scenario button properly.

We don't have the map previews/descriptions yet, but this still makes it fairly easy to open a BIQ file.

QuintillusCFC committed:
Merge branch 'Development' into BIQFilesOpenable

WildWeazel commented on C7GameData/ImportCiv3.cs:
```diff
@@ -68,11 +61,65 @@ static public C7SaveFormat ImportSav(string savePath, string defaultBicPath)
+					BaseTerrainFileID = civ3Tile.TextureFile,
+					BaseTerrainImageID = civ3Tile.TextureLocation,
+				};
+				int y = i / (theBiq.Wmap[0].Width / 2);
+				int x = (i % (theBiq.Wmap[0].Width / 2)) * 2 + (y % 2);
```
is one of these supposed to use Height? same in the method above.

QuintillusCFC replied:
Nope, that's correct.  The x uses modulus instead of division, so e.g. if i = 73 and width = 100, it will be 73 % 50 = 23.  Then times 2 = 46.  Plus y %2 = 1, X = 47, Y = 1 would be the 73rd tile (base zero) in the list.

Definitely not intuitive with the isometric grid, but it appears to be correct (other than using BIQ data for the width, it's actually copied from code Lucien wrote in December).

It probably does make sense to have it be shared across the two areas though; I'm adding a shared function now.

QuintillusCFC committed:
Put the tile-coordinate-getter code into its own helper function since it's fairly complex and shared across two places.

QuintillusCFC reviewed (commented):
Self-reviewed as this is the longest-open non-approved PR at 7 days, and changes have already been made in response to @WildWeazel 's comments.  Noticed that the helper methods in ImportCiv3.cs are inconsistent in whether they take the BIQ or the C7SaveFormat first, so I'm going to standardize them as Civ3 (source) first, C7 (destination) second.  Other than that, I think it's ready to go, and I'm looking forward to being able to open BIQs in Development.

QuintillusCFC committed:
Standardize the format of helper methods in ImportCiv3.cs as source first, destination second.
